dxil_converter.cpp

```diff
@@ -6874,6 +6883,65 @@ void Converter::Impl::emit_patch_output_lowering(CFGNode *bb)
 	}
 }
 
+void Converter::Impl::emit_clamp_tess_factor_array(spv::Id array_id, spv::Id max_tess_factor_id, uint32_t array_length)
+{
+	if (!array_length)
+		return;
+
+	auto f32_type_id = builder().makeFloatType(32);
+	auto ptr_type_id = builder().makePointer(spv::StorageClassOutput, f32_type_id);
+
+	if (!glsl_std450_ext)
+		glsl_std450_ext = builder().import("GLSL.std.450");
+
+	for (uint32_t i = 0; i < array_length; i++)
+	{
+		auto *chain_op = allocate(spv::OpInBoundsAccessChain, ptr_type_id);
+		chain_op->add_id(array_id);
+		chain_op->add_id(builder().makeUintConstant(i));
+		add(chain_op);
+
+		auto *load_op = allocate(spv::OpLoad, f32_type_id);
+		load_op->add_id(chain_op->id);
+		add(load_op);
+
+		auto *min_op = allocate(spv::OpExtInst, f32_type_id);
+		min_op->add_id(glsl_std450_ext);
+		min_op->add_literal(GLSLstd450::GLSLstd450NMin);
+		min_op->add_id(load_op->id);
+		min_op->add_id(max_tess_factor_id);
+		add(min_op);
+
+		auto *store_op = allocate(spv::OpStore);
+		store_op->add_id(chain_op->id);
+		store_op->add_id(min_op->id);
+		add(store_op);
+	}
+}
+
+void Converter::Impl::emit_clamp_tess_factors(CFGNode *node)
```

Comment:
I was expecting an implementation where the clamp happens in emit_store_patch_constant_instruction.  Fortunately, once DXIL has stored an output, it cannot be read back, so it's safe to do this.

The main complication is to check if we're storing to tess factors or not, but I think should be easy enough with has/get_builtin_shader_input and comparing the var_id.

Reply:
ok, I wasn't aware that readback isn't possible.

Reply:
Or hm ... Actually, let me double check that assumption.

Reply:
Ye, seems like my assumption is correct, nvm.

Reply:
Moved it to store_patch_constant now.